File: daytwelve/Day12.py
from collections import defaultdict
adj = defaultdict(list)
ans = 0
cur = 0
mp = defaultdict(int)

def bad_path(path):
    return path == 'start,b,d,b,A,c,A,' or path == 'start,b,d,b,A,' or path == 'start,A,c,A,c,A,b,'

def dfs(node: str, vis=dict(), twice=False, path=''):
    if ord(node[0]) >= 97 and node in vis and vis[node] == 2:
        return
    if (node == 'start' or node == 'end') and node in vis:
        return
    if twice and ord(node[0]) >= 97 and node in vis:
        return
    path += node + ','
    

    if node in vis: vis[node] += 1
    else: vis[node] = 1

    if vis[node] == 2 and ord(node[0]) >= 97:
        twice = True
    if node == 'end':
        #print(path)
        global ans
        ans += 1
        return
    for neighbor in adj[node]:
        if ord(neighbor[0]) >= 97 and neighbor in vis and vis[neighbor] == 1:
            if twice:
                continue
            dfs(neighbor, dict(vis), True, path)
        dfs(neighbor, dict(vis), twice, path)

lines = open('Day12.in').read().split('\n')
for line in lines:
    tmp = line.split('-')
    if len(tmp) != 2:
        continue
    if tmp[0] not in mp:
        mp[tmp[0]] = cur
        cur += 1
    if tmp[1] not in mp:
        mp[tmp[1]] = cur
        cur += 1
    adj[tmp[0]].append(tmp[1])
    adj[tmp[1]].append(tmp[0])

dfs("start", dict(), False, '')
print(mp)
print(ans)

File: daytwelve/Day12.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Linq;

public class Day12 {
    static long ans = 0;
    static Dictionary<string, List<string>> adj = new Dictionary<string, List<string>>();
    static int n = 12;
    static Dictionary<string, int> mp = new Dictionary<string, int>();


    static void dfs(string node, Dictionary<string, int> vis, bool twice) {
        if (node[0] >= 97 && vis.ContainsKey(node) && vis[node] == 2) return;
        if ((node == "start" || node == "end") && vis.ContainsKey(node)) return;
        if (twice && node[0] >= 97 && vis.ContainsKey(node)) return;

        if (vis.ContainsKey(node)) vis[node]++;
        else vis[node] = 1;

        if (vis[node] == 2 && node[0] >= 97) twice = true;
        if (vis[node] > 2 && node[0] >= 97) {
            Console.WriteLine(node);
            throw new Exception();
        }
        if (node == "end") {
            //foreach (string s in path) Console.Write(s + ',');
            //Console.WriteLine();
            ++ans;
            return;
        }
        foreach (string neighbor in adj[node]) {
            if (neighbor[0] >= 97 && vis.ContainsKey(neighbor) && vis[neighbor] == 1) {
                if (twice) continue;
                else {
                    dfs(neighbor, new Dictionary<string, int>(vis), true);
                }
            }
            dfs(neighbor, new Dictionary<string, int>(vis), twice);
        }
    }

    static void Main(string[] args) {
        string[][] lines = File.ReadAllLines("Day12.in").Select((string x) => x.Split('-')).ToArray();
        int cur = 0;
        foreach (string[] line in lines) {
            if (!mp.ContainsKey(line[0])) {
                mp[line[0]] = cur++;
            }
            if (!mp.ContainsKey(line[1])) {
                mp[line[1]] = cur++;
            }
            if (adj.ContainsKey(line[0])) {
                adj[line[0]].Add(line[1]);
            } else {
                adj[line[0]] = new List<string>();
                adj[line[0]].Add(line[1]);
            }
            if (adj.ContainsKey(line[1])) {
                adj[line[1]].Add(line[0]);
            } else {
                adj[line[1]] = new List<string>();
                adj[line[1]].Add(line[0]);
            }
        }
        dfs("start", new Dictionary<string, int>(), false);
        Console.WriteLine(ans);
    }
}
